### DISCUSSION TITLE
ChatGPT Prompt Template

### DISCUSSION THREAD

**Discussion by jarjee** [SilentVoid13/Templater#1046] [Templates Showcase]
First time using this plugin, thanks for making the documentation & community so good!

Now that [ChatGPT has opened up API access](https://openai.com/blog/introducing-chatgpt-and-whisper-apis), it seemed prudent to write a script & template to go with it.

```javascript
// queryChatGPT.js

async function queryChatGPT(content) {
  const apiEndpoint = "https://api.openai.com/v1/chat/completions";
  const apiKey = "<Your API Key Goes Here>";
  const authHeader = `Bearer ${apiKey}`;
  const contentTypeHeader = "application/json";

  const messages =
    typeof content === "string" ? [{ role: "user", content }] : content;

  const apiPayload = {
    model: "gpt-4o",
    messages,
  };

  const response = await fetch(apiEndpoint, {
    method: "POST",
    headers: {
      Authorization: authHeader,
      "Content-Type": contentTypeHeader,
    },
    body: JSON.stringify(apiPayload),
  });

  return await response.json();
}

module.exports = queryChatGPT;
```

And a quick template to actually call it on selected text

```markdown
<%*

const selection = tp.file.selection();

const json = await tp.user.queryChatGPT(`Answer the following in markdown.\n user: ${selection}`);

const content = json.choices[0].message.content;

return `${selection}\n${content.trim().replace(/</g, "\<").replace(/>/g, "\>")}`
%>
```

You can also pass in the "system" & "user" context arrays, the format of the input & the output from the API is [in their documentation](https://platform.openai.com/docs/guides/chat).

One thing I'd like to do, but haven't dug into yet, is getting the API key via some environment variable. I don't really like hardcoding the secret in the file.

**Comment by welpdx** [comment]
fyi: if you receive an error 429 it could be too many requests but i could also be that you: 
- reached some rate limit
- you exceeded your quota (ie your free credits ran out)
- your free credits got expired

Apparently they give you free credits that expires after 3 months. I created my account last year and I used it once lol but now they expired 🤦‍♂️

**Comment by tiagozip** [comment]
Thanks for this, I'm making a remix for a version that doesn't need API key (for free)

**Reply by geck0n** [reply]
wow! can you share it?

**Comment by threatseeker** [comment]
I would like to use that but never used scripts in obsidian and but do have chatGPT4 I like to take advantage in my notes.  Any advice on where to start?  I am no total n00b but scripts are new to me. I got the scripts folder and the script there (+ api key) and the template, but using it is not clear to me how that works.

**Reply by mtxnk** [reply]
Oh boy, i hear you

**Comment by PrFlavioCosta** [comment]
I couldn't make it work.

**Reply by andre-morise** [reply]
If you are seeing this in the Obsidian console (`cmd + option + i` for Mac and `ctrl + shift + i` for Windows), make sure your OpenAI [Credit Balance](https://platform.openai.com/account/billing/overview) is greater than $0.00 or that you have enabled "Auto Recharge." Secondly, in my experience, give it a few minutes to recognize changes to your billing. The script should start functioning correctly.
```json
▶ POST https://api.openai.com/v1/chat/completions 429 (Too Many Requests)
{
    "error": {
        "message": "You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.",
        "type": "insufficient_quota",
        "param": null,
        "code": "insufficient_quota"
    }
}
```

**Comment by conceptia2020** [comment]
Long time Obsidian user but Obsidian/Templater scripts noob (so please forgive any ignorance.)
I've created the queryChatGPT.js file, configured Templater to look for the correct folder where the js file is stored (it shows the script file as "detected.")
However, it seems to break templater's ability to parse any and all templates, even those not using this script.
For example, for testing I trimmed down my DailyNote template so that it just has the basic created / modified header template syntax (see below.)
It works fine ordinarily.
But with this script detected, I get an Template Parsing "invalid token" error.
I tested this by simply pointing the User Scripts location to a folder with no scripts in it, and suddenly my daily notes template works again.
I also tried closing/restarting Obsidian to see if that would help to no avail.
I checked the developer console, and while I have some programming knowledge, I wasn't able to understand the call stack well enough to see exactly what was triggering the error. I could only see the function call that was displaying the error.
Is this script meant to be used / compatible with the Templater plugin?
Can anyone suggest other things I can try?
I know it's me and not the script. :-)
 
```
---
created: <% tp.file.creation_date("dddd Do MMMM YYYY HH:mm A") %>
modified: <% tp.file.last_modified_date("dddd Do MMMM YYYY HH:mm A") %>
Type: DailyNote
---
```

**Reply by jarjee** [reply]
So how I've got it laid out for me is:

```
Meta
   \_ Templates
        \_ example.md
   |_ Scripts
         \_ queryChatGPT.js
```

Where I've set:
- Template folder location to `Meta/Templates` 
- Scripts folder location to `Meta/Scripts`
